File: shared/src/commonMain/kotlin/flagent/evaluator/RolloutAlgorithm.kt
package flagent.evaluator

/**
 * Canonical rollout algorithm for consistent hashing and variant selection.
 * All clients (backend, SDKs) MUST use this for identical results.
 *
 * Spec: docs/architecture/evaluation-spec.md
 * - Hash input: salt + entityID (no separator)
 * - CRC32 polynomial 0xEDB88320, bucket = crc32 % 1000
 * - Rollout: entity in rollout iff bucket < rolloutPercent * 10
 * - Distribution: accumulated percents (0-1000), binary search by bucket+1
 */
object RolloutAlgorithm {

    const val TOTAL_BUCKET_NUM = 1000u

    /**
     * CRC32 (polynomial 0xEDB88320). Pure Kotlin, no JVM-specific APIs.
     */
    fun crc32(bytes: ByteArray): UInt {
        var crc = 0xFFFFFFFFu
        for (byte in bytes) {
            crc = crc xor byte.toUByte().toUInt()
            for (k in 0 until 8) {
                crc = if ((crc and 1u) != 0u) {
                    (crc shr 1) xor 0xEDB88320u
                } else {
                    crc shr 1
                }
            }
        }
        return crc xor 0xFFFFFFFFu
    }

    /**
     * Compute bucket (0..999) for entity. Hash input = salt + entityID.
     */
    fun bucket(entityID: String, salt: String): UInt {
        val input = salt + entityID
        return crc32(input.encodeToByteArray()) % TOTAL_BUCKET_NUM
    }

    /**
     * Rollout and variant selection. Returns (variantID, debugMessage).
     * - variantIds: list of variant IDs in distribution order
     * - percentsAccumulated: accumulated percents in 0-1000 scale (e.g. [500, 1000] for 50/50)
     */
    fun rollout(
        entityID: String,
        salt: String,
        rolloutPercent: Int,
        variantIds: List<Int>,
        percentsAccumulated: List<Int>
    ): Pair<Int?, String> {
        if (entityID.isEmpty()) {
            return null to "rollout no. empty entityID"
        }
        if (rolloutPercent <= 0) {
            return null to "rollout no. invalid rolloutPercent: $rolloutPercent"
        }
        if (variantIds.isEmpty() || percentsAccumulated.isEmpty()) {
            return null to "rollout no. there's no distribution set"
        }

        val bucketNum = bucket(entityID, salt)
        val bucketInt = (bucketNum + 1u).toInt() // 1..1000

        val rolloutThreshold = rolloutPercent * 10 // 0-1000
        if (rolloutPercent < 100 && bucketInt > rolloutThreshold) {
            return null to "rollout no. entityID bucket: $bucketNum rolloutPercent: $rolloutPercent"
        }

        val index = bucketByNum(bucketInt, percentsAccumulated)
        val variantID = variantIds.getOrElse(index) { variantIds.last() }
        return variantID to "rollout yes. BucketNum: $bucketNum, VariantID: $variantID, RolloutPercent: $rolloutPercent"
    }

    /**
     * Binary search: find smallest i such that percentsAccumulated[i] >= bucketInt.
     * percentsAccumulated is in 0-1000 scale (e.g. [500, 1000]).
     */
    private fun bucketByNum(bucketInt: Int, percentsAccumulated: List<Int>): Int {
        val idx = percentsAccumulated.binarySearch(bucketInt)
        return if (idx >= 0) idx else -idx - 1
    }
}

File: sdk/swift-enhanced/Sources/FlagentEnhanced/Offline/RolloutAlgorithm.swift
import Foundation

/// Canonical rollout: salt+entityID, CRC32, bucket 0..999. Matches shared/evaluator/RolloutAlgorithm.kt.
enum RolloutAlgorithm {
    static let totalBucketNum: UInt32 = 1000

    /// CRC32 (polynomial 0xEDB88320).
    static func crc32(_ bytes: [UInt8]) -> UInt32 {
        var crc: UInt32 = 0xFFFF_FFFF
        for byte in bytes {
            crc ^= UInt32(byte)
            for _ in 0..<8 {
                crc = (crc & 1) != 0 ? (crc >> 1) ^ 0xEDB8_8320 : crc >> 1
            }
        }
        return crc ^ 0xFFFF_FFFF
    }

    /// Bucket 0..999. Input = salt + entityID.
    static func bucket(entityID: String, salt: String) -> UInt32 {
        let input = salt + entityID
        let data = Array(input.utf8)
        return crc32(data) % totalBucketNum
    }

    /// Returns (variantID, message). variantIds and percentsAccumulated in 0-1000 scale.
    static func rollout(
        entityID: String,
        salt: String,
        rolloutPercent: Int,
        variantIds: [Int],
        percentsAccumulated: [Int]
    ) -> (Int?, String) {
        if entityID.isEmpty {
            return (nil, "rollout no. empty entityID")
        }
        if rolloutPercent <= 0 {
            return (nil, "rollout no. invalid rolloutPercent: \(rolloutPercent)")
        }
        if variantIds.isEmpty || percentsAccumulated.isEmpty {
            return (nil, "rollout no. there's no distribution set")
        }
        let bucketNum = bucket(entityID: entityID, salt: salt)
        let bucketInt = Int(bucketNum) + 1
        let rolloutThreshold = rolloutPercent * 10
        if rolloutPercent < 100 && bucketInt > rolloutThreshold {
            return (nil, "rollout no. entityID bucket: \(bucketNum) rolloutPercent: \(rolloutPercent)")
        }
        let index = bucketByNum(bucketInt: bucketInt, percentsAccumulated: percentsAccumulated)
        let variantID = index < variantIds.count ? variantIds[index] : variantIds.last!
        return (variantID, "rollout yes. BucketNum: \(bucketNum), VariantID: \(variantID), RolloutPercent: \(rolloutPercent)")
    }

    /// Smallest i such that percentsAccumulated[i] >= bucketInt.
    private static func bucketByNum(bucketInt: Int, percentsAccumulated: [Int]) -> Int {
        var low = 0
        var high = percentsAccumulated.count
        while low < high {
            let mid = low + (high - low) / 2
            if percentsAccumulated[mid] < bucketInt {
                low = mid + 1
            } else {
                high = mid
            }
        }
        return low
    }
}
